src/Polysemy/Internal/Union.hs

```diff
@@ -232,4 +237,20 @@ decompCoerce (Union p a) =
     SS n -> Left (Union (SS n) a)
 {-# INLINE decompCoerce #-}
 
+class LastMembers end effs | effs -> end where
+  decompLast
+      :: Union effs m a
+      -> Either (Union effs m a) (Union '[end] m a)
+
+instance {-# OVERLAPPABLE #-} LastMembers end effs => LastMembers end (eff ': effs) where
+  decompLast (Union SZ u)     = Left $ Union SZ u
+  decompLast (Union (SS n) u) = first weaken $ decompLast $ Union n u
+
+instance LastMembers end (end ': '[]) where
```

Comment:
Curiosity:

Is it possible to use `'[end]` instead of `(end ': '[])` here?

Reply:
Probably!